src/IO/ReadBufferFromRemoteFS.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include <Disks/IDiskRemote.h>
+#include <IO/ReadBufferFromFile.h>
+
+
+namespace DB
+{
+
+class ReadBufferFromRemoteFS : public ReadBufferFromFileBase
+{
+friend class ThreadPoolRemoteFSReader;
+
+public:
+    explicit ReadBufferFromRemoteFS(const RemoteMetadata & metadata_);
+
+    off_t seek(off_t offset, int whence) override;
+
+    off_t getPosition() override { return absolute_position - available(); }
+
+    String getFileName() const override { return metadata.metadata_file_path; }
+
+    void reset();
+
+protected:
+    size_t readInto(char * data, size_t size, size_t offset);
+
+    virtual SeekableReadBufferPtr createReadBuffer(const String & path) const = 0;
+
+    RemoteMetadata metadata;
+
+private:
+    bool nextImpl() override;
+
+    SeekableReadBufferPtr initialize();
+
+    bool read();
```

Comment:
We already have a lot of `read` methods in ReadBuffer base class. This name confusing.

Reply:
a better name would be `readImpl`, but since there is already `nextImpl`  it is more easier to confuse them with each other. So I named it just `read`, can you suggest a better name? I have no more fantasy.